File: sandrose/hr/report/employees_on_leave_as_on_date/employees_on_leave_as_on_date.py
import frappe
from frappe import _
from frappe.utils import getdate, date_diff


def execute(filters=None):
    if not filters:
        filters = {}

    as_on_date = getdate(filters.get("as_on_date") or frappe.utils.today())

    columns = get_columns()
    data = get_data(as_on_date)
    return columns, data


def get_columns():
    return [
        {"label": _("Code"),        "fieldname": "employee",       "fieldtype": "Link", "options": "Employee", "width": 110},
        {"label": _("Name"),        "fieldname": "employee_name",  "fieldtype": "Data", "width": 180},
        {"label": _("Designation"), "fieldname": "designation",    "fieldtype": "Data", "width": 150},
        {"label": _("Department"),  "fieldname": "department",     "fieldtype": "Data", "width": 180},
        {"label": _("Category"),    "fieldname": "employment_type","fieldtype": "Data", "width": 100},
        {"label": _("Leave Type"),  "fieldname": "leave_type",     "fieldtype": "Data", "width": 140},
        {"label": _("Leave From"),  "fieldname": "from_date",      "fieldtype": "Date", "width": 100},
        {"label": _("Leave To"),    "fieldname": "to_date",        "fieldtype": "Date", "width": 100},
        {"label": _("Rejoined"),    "fieldname": "rejoined",       "fieldtype": "Date", "width": 100},
        {"label": _("Allowed"),     "fieldname": "allowed",        "fieldtype": "Int",  "width": 80},
        {"label": _("Excess"),      "fieldname": "excess",         "fieldtype": "Int",  "width": 80},
        {"label": _("Contact"),     "fieldname": "cell_number",    "fieldtype": "Data", "width": 130},
    ]


def get_data(as_on_date):
    leaves = frappe.db.sql("""
        SELECT
            la.employee,
            la.employee_name,
            la.leave_type,
            la.from_date,
            la.to_date,
            la.total_leave_days,
            la.department,
            e.designation,
            e.employment_type,
            e.cell_number,
            e.custom_rejoining
        FROM `tabLeave Application` la
        INNER JOIN `tabEmployee` e ON e.name = la.employee
        WHERE
            la.docstatus = 1
            AND la.status = 'Approved'
            AND la.from_date <= %(as_on_date)s
            AND la.to_date >= %(as_on_date)s
        ORDER BY la.department, la.leave_type, la.employee
    """, {"as_on_date": as_on_date}, as_dict=True)

    data = []
    for row in leaves:
        allowed = get_leave_allocation(row.employee, row.leave_type, as_on_date)

        taken = row.total_leave_days or 0
        excess = int(taken - allowed) if (allowed is not None and taken > allowed) else None

        rejoined = None
        if row.custom_rejoining and getdate(row.to_date) < as_on_date:
            rejoined = row.custom_rejoining

        data.append({
            "employee":        row.employee,
            "employee_name":   row.employee_name,
            "designation":     row.designation,
            "department":      row.department,
            "employment_type": row.employment_type,
            "leave_type":      row.leave_type,
            "from_date":       row.from_date,
            "to_date":         row.to_date,
            "rejoined":        rejoined,
            "allowed":         int(allowed) if allowed is not None else 0,
            "excess":          excess,
            "cell_number":     row.cell_number,
        })

    return data


def get_leave_allocation(employee, leave_type, as_on_date):
    result = frappe.db.sql("""
        SELECT total_leaves_allocated
        FROM `tabLeave Allocation`
        WHERE
            employee   = %(employee)s
            AND leave_type = %(leave_type)s
            AND docstatus  = 1
            AND from_date  <= %(as_on_date)s
            AND to_date    >= %(as_on_date)s
        ORDER BY from_date DESC
        LIMIT 1
    """, {
        "employee":   employee,
        "leave_type": leave_type,
        "as_on_date": as_on_date,
    })

    return result[0][0] if result else None

File: sandrose/sandrose/report/employees_on_leave_as_on_date/employees_on_leave_as_on_date.js
frappe.query_reports["Employees On Leave As On Date"] = {
filters: [
{
fieldname: "as_on_date",
label: __("As On Date"),
fieldtype: "Date",
default: frappe.datetime.get_today(),
reqd: 1,
},
{
fieldname: "company",
label: __("Company"),
fieldtype: "Link",
options: "Company",
},
],

onload: function (report) {
report.page.add_inner_button(__("Print"), function () {
let filters = report.get_values();
if (!filters) return;

let data = report.data;
if (!data || data.length === 0) {
frappe.msgprint(__("No data to print."));
return;
}

let company = "SANDROSE TRADING LLC";
let as_on_date = frappe.datetime.str_to_user(filters.as_on_date);
let print_date = frappe.datetime.str_to_user(frappe.datetime.get_today());

let grouped = {};
data.forEach((row) => {
let dept = row.department || "\u2014";
let leave_type = row.leave_type || "\u2014";
if (!grouped[dept]) grouped[dept] = {};
if (!grouped[dept][leave_type]) grouped[dept][leave_type] = [];
grouped[dept][leave_type].push(row);
});

let rows_html = "";
Object.keys(grouped).forEach((dept) => {
rows_html += "<tr><td colspan='9' style='font-weight:bold; padding: 6px 4px 2px 4px; font-size: 11pt;'>Department : " + dept + "</td></tr>";

Object.keys(grouped[dept]).forEach((leave_type) => {
rows_html += "<tr><td colspan='9' style='padding: 2px 4px 2px 20px; font-style: italic; font-size: 10.5pt; font-weight: 600;'>" + leave_type + "</td></tr>";

grouped[dept][leave_type].forEach((row) => {
rows_html += "<tr>"
+ "<td>" + (row.employee || "") + "</td>"
+ "<td>" + (row.employee_name || "") + "</td>"
+ "<td>" + (row.designation || "") + "</td>"
+ "<td>" + (row.employment_type || row.category || "") + "</td>"
+ "<td>" + (row.from_date ? frappe.datetime.str_to_user(row.from_date) : "") + "</td>"
+ "<td>" + (row.to_date ? frappe.datetime.str_to_user(row.to_date) : "") + "</td>"
+ "<td>" + (row.rejoined || "") + "</td>"
+ "<td style='text-align:center;'>" + (row.allowed || "") + "</td>"
+ "<td style='text-align:center;'>" + (row.excess || "") + "</td>"
+ "</tr>";
});
});
});

let html = "<!DOCTYPE html><html><head><meta charset='UTF-8'>"
+ "<title>Employees On Leave - " + company + "</title>"
+ "<style>"
+ "* { margin: 0; padding: 0; box-sizing: border-box; }"
+ "body { font-family: Arial, sans-serif; font-size: 10pt; color: #000; background: #fff; }"
+ ".page-wrapper { width: 100%; max-width: 900px; margin: 0 auto; padding: 24px 32px; }"
+ ".report-header { text-align: center; margin-bottom: 10px; border-bottom: 2px solid #000; padding-bottom: 8px; }"
+ ".company-name { font-size: 16pt; font-weight: bold; letter-spacing: 0.5px; }"
+ ".report-title { font-size: 11pt; font-weight: bold; margin-top: 2px; }"
+ ".meta-row { display: flex; justify-content: space-between; margin: 6px 0 10px 0; font-size: 9.5pt; }"
+ "table { width: 100%; border-collapse: collapse; font-size: 9.5pt; }"
+ "thead tr th { border-top: 1.5px solid #000; border-bottom: 1.5px solid #000; padding: 5px 4px; text-align: left; font-weight: bold; white-space: nowrap; }"
+ "tbody tr td { padding: 3px 4px; vertical-align: top; white-space: nowrap; }"
+ "tbody tr:last-child td { border-bottom: 1.5px solid #000; }"
+ "@media print { .no-print { display: none !important; } @page { size: landscape; margin: 10mm 12mm; } }"
+ "</style></head><body>"
+ "<div class='page-wrapper'>"
+ "<div class='no-print' style='text-align:right; margin-bottom:12px;'>"
+ "<button onclick='window.print()' style='padding:6px 18px; font-size:10pt; cursor:pointer; background:#171717; color:#fff; border:none; border-radius:3px;'>Print</button>"
+ "</div>"
+ "<div class='report-header'>"
+ "<div class='company-name'>" + company.toUpperCase() + "</div>"
+ "<div class='report-title'>Employees On Leave As On " + as_on_date + "</div>"
+ "</div>"
+ "<div class='meta-row'><span>" + print_date + "</span><span>Page 1 of 1</span></div>"
+ "<table><thead><tr>"
+ "<th>Code</th><th>Name</th><th>Designation</th><th>Category</th>"
+ "<th>Leave From</th><th>Leave To</th><th>Rejoined</th><th>Allowed</th><th>Excess</th>"
+ "</tr></thead><tbody>" + rows_html + "</tbody></table>"
+ "</div></body></html>";

let w = window.open("", "_blank", "width=960,height=700");
w.document.write(html);
w.document.close();
});
},
};
